server/src/main/java/com/exacaster/lighter/application/sessions/processors/python/PythonSessionIntegration.java

```diff
@@ -95,7 +95,7 @@ public void runServer(StartupEvent event) {
 
     @Override
     public boolean hasWaitingStatement(Application application) {
-        var waitingStatements = statementStorage.findByState(application.getAppId(), "waiting");
```

Comment:
Basically this was the root cause.
Other changes are refactoring - decoupling status tracker from sessions layer.